File: frontend/src/locationProvider.js
// Centralized Location Provider for the UI
// Coordinates prioritize:
// 1. Live GPS coordinates (if available and valid)
// 2. Current Device Location (via browser geolocation prompt on startup)
// 3. Last Known Location (loaded from localStorage cache)
// 4. Configured Demo Coordinates (fallback, via VITE_DEMO_LATITUDE / VITE_DEMO_LONGITUDE env variables)
// 5. Existing hardcoded values (last resort)

import { useState, useEffect } from 'react';

export const DEFAULT_LATITUDE = 12.9716;
export const DEFAULT_LONGITUDE = 77.5946;

// Internal central location coordinates state and source
let centralLatitude = null;
let centralLongitude = null;
let centralSource = 'Default Fallback Coordinates';

// Cache key for persisting geolocation
const CACHE_KEY = 'water_monitor_last_known_location';

// Registry of listeners for location update events
const listeners = [];

/**
 * Returns resolved coordinates and source.
 */
export function getActiveLocationWithSource(gpsLat, gpsLon) {
  // Priority 1: GPS Coordinates when available and valid (real device overrides everything)
  if (gpsLat !== undefined && gpsLat !== null && !isNaN(gpsLat) && Math.abs(gpsLat) > 0.0001) {
    if (gpsLon !== undefined && gpsLon !== null && !isNaN(gpsLon) && Math.abs(gpsLon) > 0.0001) {
      return [parseFloat(gpsLat), parseFloat(gpsLon), 'GPS'];
    }
  }

  // Priority 2 / 3: Geolocation detected on startup or loaded from cache
  if (centralLatitude !== null && centralLongitude !== null) {
    return [centralLatitude, centralLongitude, centralSource];
  }

  // Priority 4: Configured Demo Coordinates (from Vite environment variables)
  const configLat = typeof import.meta !== 'undefined' && import.meta.env && import.meta.env.VITE_DEMO_LATITUDE
    ? parseFloat(import.meta.env.VITE_DEMO_LATITUDE)
    : null;
  const configLon = typeof import.meta !== 'undefined' && import.meta.env && import.meta.env.VITE_DEMO_LONGITUDE
    ? parseFloat(import.meta.env.VITE_DEMO_LONGITUDE)
    : null;

  if (configLat !== null && !isNaN(configLat) && configLon !== null && !isNaN(configLon)) {
    return [configLat, configLon, 'Demo Environment Variables'];
  }

  // Priority 5: Existing hardcoded values (last resort)
  return [DEFAULT_LATITUDE, DEFAULT_LONGITUDE, 'Default Fallback Coordinates'];
}

/**
 * Returns resolved coordinates only.
 */
export function getActiveLocation(gpsLat, gpsLon) {
  const [lat, lon] = getActiveLocationWithSource(gpsLat, gpsLon);
  return [lat, lon];
}

/**
 * Subscribe a component or data module to location update events.
 * Returns an unsubscribe cleanup function.
 */
export function subscribeToLocation(callback) {
  listeners.push(callback);
  
  // Call immediately with the current state to initialize
  const [lat, lon, source] = getActiveLocationWithSource();
  callback(lat, lon, source);
  
  return () => {
    const idx = listeners.indexOf(callback);
    if (idx !== -1) {
      listeners.splice(idx, 1);
    }
  };
}

/**
 * Updates the central location state, updates caches, notifies listeners,
 * and synchronizes the location to the backend API.
 */
export function updateCentralLocation(lat, lon, source) {
  const lat_f = parseFloat(lat);
  const lon_f = parseFloat(lon);
  
  if (isNaN(lat_f) || isNaN(lon_f)) return;
  
  centralLatitude = lat_f;
  centralLongitude = lon_f;
  centralSource = source;
  
  console.log(`[LocationProvider] Active coordinates set to (${lat_f.toFixed(6)}, ${lon_f.toFixed(6)}) via ${source}`);

  // Notify all subscribed listeners
  listeners.forEach(callback => {
    try {
      callback(lat_f, lon_f, source);
    } catch (e) {
      console.error("Error executing location listener callback:", e);
    }
  });
}

/**
 * Synchronize location update to backend Fast-API server so simulator coordinates align.
 */
async function syncLocationToBackend(lat, lon, source) {
  try {
    const resp = await fetch('http://localhost:8000/location', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        latitude: lat,
        longitude: lon,
        source: source
      })
    });
    const result = await resp.json();
    console.log('[LocationProvider] Synchronised with backend:', result);
  } catch (err) {
    console.warn('[LocationProvider] Synchronisation to backend failed (backend may be starting up):', err.message);
  }
}

/**
 * React state hook for coordinates. Auto-updates when location is detected.
 */
export function useActiveLocation(gpsLat, gpsLon) {
  const [locationState, setLocationState] = useState(() => {
    return getActiveLocationWithSource(gpsLat, gpsLon);
  });

  useEffect(() => {
    // If a valid live GPS override is provided, bypass central subscription updates
    if (gpsLat !== undefined && gpsLat !== null && !isNaN(gpsLat) && Math.abs(gpsLat) > 0.0001) {
      if (gpsLon !== undefined && gpsLon !== null && !isNaN(gpsLon) && Math.abs(gpsLon) > 0.0001) {
        setLocationState([parseFloat(gpsLat), parseFloat(gpsLon), 'GPS']);
        return;
      }
    }

    // Subscribe to central updates (triggered by device detection, updates, or cache fetches)
    return subscribeToLocation((lat, lon, source) => {
      setLocationState([lat, lon, source]);
    });
  }, [gpsLat, gpsLon]);

  return locationState;
}

// ── Startup Location Detection ───────────────────────────────────────
if (typeof window !== 'undefined') {
  // A. First try to load cached last known coordinates to initialize instantly
  try {
    const cached = window.localStorage.getItem(CACHE_KEY);
    if (cached) {
      const [lat, lon] = JSON.parse(cached);
      updateCentralLocation(lat, lon, 'Last Known Location');
      // Sync cached location to backend in case it restarted
      syncLocationToBackend(lat, lon, 'Last Known Location');
    }
  } catch (e) {
    console.warn('[LocationProvider] Cache read failed:', e);
  }

  // B. Trigger asynchronous browser geolocation prompt
  if (navigator.geolocation) {
    console.log('[LocationProvider] Initiating browser Geolocation detection...');
    navigator.geolocation.getCurrentPosition(
      (position) => {
        const lat = position.coords.latitude;
        const lon = position.coords.longitude;
        
        console.log(`[LocationProvider] Browser Geolocation success: (${lat.toFixed(6)}, ${lon.toFixed(6)})`);
        
        // Cache locally in localStorage
        try {
          window.localStorage.setItem(CACHE_KEY, JSON.stringify([lat, lon]));
        } catch (e) {
          console.warn('[LocationProvider] Cache write failed:', e);
        }
        
        // Update provider and notify app
        updateCentralLocation(lat, lon, 'Browser Geolocation');
        // Synchronise browser coordinates to FastAPI backend simulator
        syncLocationToBackend(lat, lon, 'Browser Geolocation');
      },
      (error) => {
        console.warn(`[LocationProvider] Geolocation lookup failed: ${error.message} (Code: ${error.code})`);
      },
      { enableHighAccuracy: true, timeout: 5000, maximumAge: 0 }
    );
  } else {
    console.warn('[LocationProvider] Browser navigator.geolocation is not supported on this environment.');
  }
}

File: utils/location_provider.py
"""
location_provider.py — Centralized active-location provider.

Configured coordinates are prioritized according to the following rules:
  1. GPS Coordinates (when available and valid)
  2. Current Device Location (via IP geolocation on startup)
  3. Last Known Location (loaded from cache)
  4. Configured Demo Coordinates (via environment variables)
  5. Existing hardcoded values (last resort)
"""

from __future__ import annotations
import os
import json
import logging

logger = logging.getLogger(__name__)

class LocationProvider:
    # Predefined / Configured demo coordinates (fallback when GPS is not functional)
    # Developers can set these environment variables to change the location globally.
    DEMO_LATITUDE = float(os.getenv("DEMO_LATITUDE", os.getenv("VITE_DEMO_LATITUDE", "12.9716")))
    DEMO_LONGITUDE = float(os.getenv("DEMO_LONGITUDE", os.getenv("VITE_DEMO_LONGITUDE", "77.5946")))
    
    # Last resort hardcoded fallback coordinates
    HARDCODED_LATITUDE = 12.9716
    HARDCODED_LONGITUDE = 77.5946

    # Dynamic active location state
    _active_latitude: float | None = None
    _active_longitude: float | None = None
    _location_source: str = "Default Fallback Coordinates"
    
    # Cache file to store last known successful location
    _cache_file: str = ".last_known_location.json"
    
    # Track the last logged coordinates/source to prevent log spam in the loop
    _last_logged_resolved: tuple[float, float, str] | None = None

    @classmethod
    def detect_device_location(cls) -> None:
        """
        Attempts to obtain current device location using IP geolocation,
        falling back to last known location cache.
        """
        # 1. Attempt device IP Geolocation
        try:
            import requests
            logger.info("Attempting device IP geolocation lookup...")
            # Use ip-api.com (free, no key required). Timeout 3.0s to avoid hanging on start.
            resp = requests.get("http://ip-api.com/json/", timeout=3.0)
            if resp.status_code == 200:
                data = resp.json()
                lat = float(data.get("lat"))
                lon = float(data.get("lon"))
                if -90.0 <= lat <= 90.0 and -180.0 <= lon <= 180.0:
                    cls.update_location(lat, lon, "IP Geolocation")
                    logger.info("Current device location successfully detected: (%.6f, %.6f) via IP Geolocation", lat, lon)
                    return
        except Exception as e:
            logger.warning("IP geolocation lookup failed: %s", e)

        # 2. Fallback to Last Known Location cache
        if os.path.exists(cls._cache_file):
            try:
                with open(cls._cache_file, "r") as f:
                    cache_data = json.load(f)
                lat = float(cache_data.get("latitude"))
                lon = float(cache_data.get("longitude"))
                cls._active_latitude = lat
                cls._active_longitude = lon
                cls._location_source = "Last Known Location"
                logger.info("Loaded last known location from cache: (%.6f, %.6f) [Source: Last Known Location]", lat, lon)
                return
            except Exception as e:
                logger.warning("Failed to load last known location cache: %s", e)

        logger.info("Device geolocation and cache unavailable. Falling back to env variables / defaults.")

    @classmethod
    def update_location(cls, lat: float, lon: float, source: str) -> None:
        """
        Updates the active location coordinates and persists them to the cache.
        """
        try:
            cls._active_latitude = float(lat)
            cls._active_longitude = float(lon)
            cls._location_source = source
            
            logger.info("Active location updated to: (%.6f, %.6f) [Source: %s]", cls._active_latitude, cls._active_longitude, source)
            
            # Persist to last known cache
            cache_data = {"latitude": cls._active_latitude, "longitude": cls._active_longitude}
            with open(cls._cache_file, "w") as f:
                json.dump(cache_data, f)
        except Exception as e:
            logger.error("Failed to update location or save cache: %s", e)

    @classmethod
    def get_active_location_with_source(cls, gps_lat: float | None = None, gps_lon: float | None = None) -> tuple[float, float, str]:
        """
        Retrieves the active coordinates and the source name using the priority rules:
          1. GPS Coordinates (when available and valid)
          2. Detected Device Location (IP Geolocation on startup / Browser geolocation update)
          3. Last Known Location (loaded from cache)
          4. Configured Demo Coordinates (fallback)
          5. Existing hardcoded values (last resort)
        """
        resolved_lat = cls.HARDCODED_LATITUDE
        resolved_lon = cls.HARDCODED_LONGITUDE
        source = "Default Fallback Coordinates"

        # 1. GPS Coordinates (when available and valid)
        if gps_lat is not None and gps_lon is not None:
            try:
                gps_lat_f = float(gps_lat)
                gps_lon_f = float(gps_lon)
                if -90.0 <= gps_lat_f <= 90.0 and -180.0 <= gps_lon_f <= 180.0:
                    if abs(gps_lat_f) > 0.0001 and abs(gps_lon_f) > 0.0001:
                        resolved_lat = gps_lat_f
                        resolved_lon = gps_lon_f
                        source = "GPS"
            except (ValueError, TypeError):
                pass

        # 2/3. Device / Cached Coordinates
        elif cls._active_latitude is not None and cls._active_longitude is not None:
            resolved_lat = cls._active_latitude
            resolved_lon = cls._active_longitude
            source = cls._location_source

        # 4. Configured Demo Coordinates (explicit env variables)
        elif (os.getenv("DEMO_LATITUDE") or os.getenv("VITE_DEMO_LATITUDE") or 
              os.getenv("DEMO_LONGITUDE") or os.getenv("VITE_DEMO_LONGITUDE")):
            resolved_lat = cls.DEMO_LATITUDE
            resolved_lon = cls.DEMO_LONGITUDE
            source = "Demo Environment Variables"

        # Change-sensitive logging to prevent spamming logs
        resolved_tuple = (resolved_lat, resolved_lon, source)
        if cls._last_logged_resolved != resolved_tuple:
            logger.info("Resolved active location: (%.6f, %.6f) using source: %s", resolved_lat, resolved_lon, source)
            cls._last_logged_resolved = resolved_tuple

        return resolved_lat, resolved_lon, source

    @classmethod
    def get_active_location(cls, gps_lat: float | None = None, gps_lon: float | None = None) -> tuple[float, float]:
        """
        Compatibility wrapper that returns only (lat, lon) coordinates.
        """
        lat, lon, _ = cls.get_active_location_with_source(gps_lat, gps_lon)
        return lat, lon
